raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,298 @@
+    /// @param _rsb_address The address of the RaidenServiceBundle contract
+    constructor(
+        address _token_address,
+        address _rsb_address
+    )
+        public
+    {
+        require(_token_address != 0x0);
+        require(_rsb_address != 0x0);
+        require(contractExists(_token_address));
+        require(contractExists(_rsb_address));
+
+        token = Token(_token_address);
+        rsb = RaidenServiceBundle(_rsb_address);
+        // Check if the contract is indeed a token contract
+        require(token.totalSupply() > 0);
+        // Check if the contract is indeed an rsb contract
+        // TODO: Check that some function exists in the contract
+    }
+
+    /// @notice Deposit tokens used to either pay MSs or to regsiter as MS
+    /// Can be called by anyone several times and on behalf of other accounts
+    /// @param beneficiary The account benefitting from the deposit
+    /// @param amount The amount of tokens to be depositted
+    function deposit(address beneficiary, uint amount) public {
+        require(amount > 0);
+
+        // This also allows for MSs to deposit and use other MSs
+        balances[beneficiary] += amount;
+
+        // Transfer the deposit to the smart contract
+        require(token.transferFrom(msg.sender, address(this), amount));
+
+        emit NewDeposit(beneficiary, amount);
+    }
+
+    /// @notice Internal function that updates the Reward struct if a newer balance proof
+    /// is provided in the monitor() function
+    /// @param token_network_address Address of the TokenNetwork being monitored
+    /// @param closing_participant The address of the closing participant
+    /// @param non_closing_participant Address of the other channel participant
+    /// @param reward_amount The amount of tokens to be rewarded
+    /// @param nonce The nonce of the newly provided balance_proof
+    /// @param monitoring_service_address The address of the MS calling monitor()
+    /// @param reward_proof_signature The signature of the signed reward proof
+    function updateReward(
+        address token_network_address,
+        address closing_participant,
+        address non_closing_participant,
+        uint256 reward_amount,
+        uint256 nonce,
+        address monitoring_service_address,
+        bytes reward_proof_signature
+    )
+    internal
+    {
+        TokenNetwork token_network = TokenNetwork(token_network_address);
+        bytes32 channel_identifier = token_network.getChannelIdentifier(closing_participant, non_closing_participant);
+
+        // Make sure that the reward proof is signed by the non_closing_participant
+        address raiden_node_address = recoverAddressFromRewardProof(
+            channel_identifier,
+            reward_amount,
+            token_network_address,
+            token_network.chain_id(),
+            nonce,
+            reward_proof_signature
+        );
+        require(raiden_node_address == non_closing_participant);
```

Comment:
Doesn't need changes in this PR, but keep in mind this https://github.com/raiden-network/raiden-contracts/issues/47 - code style decision of keeping the checks grouped. If directly related to the input, they should be first. Otherwise, they can be added last. This requires attention to not change a variable between the operation that you want to check and the actual check - so there are some drawbacks...